packages/components/src/custom-select-control-v2/types.ts

```diff
@@ -40,15 +33,131 @@ export type CustomSelectProps = {
 	/**
 	 * The size of the control.
 	 *
-	 * @default 'default'
+	 * @default default
 	 */
-	size?: 'default' | 'small';
+	size?: 'compact' | 'default' | 'small';
 	/**
 	 * Can be used to externally control the value of the control.
 	 */
 	value?: string | string[];
 };
 
+export type _CustomSelectProps = {
+	/**
+	 * The child elements. This should be composed of CustomSelectItem components.
+	 */
+	children: React.ReactNode;
+	/**
+	 * Used to visually hide the label. It will always be visible to screen readers.
+	 *
+	 * @default false
+	 */
+	hideLabelFromVision?: boolean;
+	/**
+	 * Label for the control.
+	 */
+	label: string;
+};
+
+export type CustomSelectProps = _CustomSelectProps &
+	Exclude< CustomSelectButtonProps, { size: 'small' } >;
+
+/**
+ * The legacy object structure for the options array.
+ */
+type Option = {
+	key: string;
+	name: string;
+	style?: {};
+	className?: string;
+	__experimentalHint?: string;
+};
+
+/**
+ * The legacy object returned from the onChange event.
+ */
+type OnChangeObject = {
```

Comment:
Nit: Improves clarity when looking at unionized `onChange` signatures in IntelliSense.

```suggestion
type LegacyOnChangeObject = {
```